applications/NeuroZK.md

```diff
@@ -0,0 +1,266 @@
+# NeuroZK - Empowering AI at the edge with zero-knowledge proofs.
+
+- **Team Name:** Cyborg Network
+- **Payment Address:** 12Enn321pZSZkNtBwsTRxxZTKS9DsGpk8PTkZsnsMjTzrtFa (AssetHub USDT) 
+- **DOT Vesting Address:** 15p7mTm9uKt1w7GF3AU1TNPWm4eAHgieTjtfsVVEtqHazpaY
+- **Level:** 2
+
+## Project Overview :page_facing_up:
+
+![Untitled design (1)](https://github.com/Cyborg-Network/Grants-Program/assets/93442895/a33de305-fba2-4a43-ac73-fe3736c03b56)
+
+### Overview
+
+This application is a follow up from our successful [grant delivery](https://github.com/w3f/Grant-Milestone-Delivery/pull/1196)
+
+Cyborg Network is developing a DePIN-based marketplace for edge computing resources, enabling latency-sensitive applications to scale globally with ease. Users can seamlessly offload specific low-latency workloads to edge locations while maintaining their core deployment in the cloud. Our vision is to serve as essential middleware at the network edge, orchestrating these tasks to reduce cloud operational costs and enhance existing cloud services. In response to user demand and recognizing the growing importance of distributed AI inference infrastructure, we are prioritizing the development of a decentralized AI compute platform leveraging ZK proofs as our go-to-market strategy.
+
+As part of our previous application, we delivered CyberDock, a product line focused on helping users deploy dockerized containers into our edge server network. We are now focusing on expanding this approach to a new product line called NeuroZK, which will enable AI apps to run Inference Engines backed by ZK proofs.
+
+Users can easily run GPU based inference engines in our crowdsoruced GPU network, with compute integrity verified using a ZK SNARK proof that proves that the inference algorithm was executed with 100% accuracy without revealing the actual inference algorithm nor the query or result data.
```

Comment:
Apologies for the delayed response, I have updated the document to include the actual verification logic.

Here is a detailed explainer that also covers your specific questions.

**Proof Generation:**

Once a task is assigned to a particular worker node and it starts executing the task, at fixed intervals of time (currently arbitrarily set to 60 minutes), the ZK worker component in the worker node will generate a SNARK proof using the PLONK algorithm. We will use an oracle to convey this proof to the verifier pallet, which will evaluate the received proof using the verification key and public data inputs that the user submitted before execution.

**Verification:**

the worker node submits a ZK proof to the blockchain. This proof is a compact set of cryptographic commitments that represent the correctness of a computation.
The verifier pallet  uses public data (submitted earlier or as part of the transaction) as inputs for the verification.
The verifier pallet uses a verification key, generated during the trusted setup phase, to check the proof. This key is static and is uploaded during task deployment. The pallet performs a series of elliptic curve operations and field arithmetic (over finite fields) to check the polynomial commitments and evaluations provided in the proof. The commitments must match the expected public inputs and satisfy the polynomial equations set by the PLONK circuit.Permutation arguments and other constraints encoded into the circuit are also validated at this stage. The Pallet checks whether all evaluations and commitments hold for the circuit, using the verification key. If they match, the proof is accepted as valid; otherwise, it is rejected. Once the verification succeeds, the transaction is processed as valid (e.g., approving the computation result). If the verification fails, the transaction is reverted or flagged as invalid.

**Computational costs related to ZKP on Cyborg:**

Since proof generation occurs within Cyborg worker nodes—custom spec edge servers designed for executing LLMs and sold by our organization—the costs related to proof generation can be neglected, as the compute providers are compensated for executing inference tasks rather than for proof generation. Regarding verification costs, PLONK proof sizes are typically around 100 bytes; however, the actual gas fees will depend on the number of constraints involved in the elliptic curve operations. As this is still experimental, we cannot currently estimate the gas costs for verification theoretically. However, after successful implementation of the system, we can optimize verification gas by performing batch verification of proofs aggregated off-chain.